feat(curriculum): add grouping characters challenge and guide (#34875)
freeCodeCamp/freeCodeCamp: curriculum/challenges/english/02-javascript-algorithms-and-data-structures/regular-expressions/check-for-mixed-grouping-of-characters.english.md

---
id: 5c3dda8b4d8df89bea71600f
title: Check For Mixed Grouping of Characters
challengeType: 1
---

## Description
<section id='description'>

> Sometimes we want to check for groups of characters using a Regular Expression, to achieve this task we use the round brackets <code>()</code>.

Reviewer: ```suggestion
Sometimes we want to check for groups of characters using a Regular Expression and to achieve that we can use parentheses <code>()</code>.
```

> If you want to find either <code>"Penguin"</code> or <code>"Pumpkin"</code> in a string, you could use the following Regular Expression: <code>/P(engu|umpk)in/g</code>

Reviewer: ```suggestion
If you want to find either <code>Penguin</code> or <code>Pumpkin</code> in a string, you can use the following Regular Expression: <code>/P(engu|umpk)in/g</code>
```

> Then you can test the code using the <code>test()</code> method to find whether the requires character(s)/string(s) are in the test string or not.

Reviewer: ```suggestion
Then check whether the desired string groups are in the test string by using the <code>test()</code> method.
```

Author: Do we want to write string groups or just strings?

<blockquote>let testStr = "Pumpkin";<br>let testRegex = /P(engu|umpk)in/g;<br>testRegex.test(testStr);<br>// Returns true</blockquote>
</section>

## Instructions
<section id='instructions'>

> Fix the RegEx so that it checks for either <code>Hello</code> or <code>Halo</code> in a case sensitive manner.

Reviewer: RegEx: pick one spelling and stick to it. The most common in this exercise has been without abbreviation, but do check what the ones before and after use.

Author: I'll change everything to regex, seems like the most used in the curriculum

</section>

## Tests
<section id='tests'>

```yml
tests:
  - text: Your regex should return <code>true</code> for the string <code>"Hello"</code>.
    testString: assert(myRegex.test('Hello'), 'Your regex should return <code>true</code> for the string <code>"Hello"</code>.');
  - text: Your regex should return <code>true</code> for the string <code>"Halo"</code>.
    testString: assert(myRegex.test('Halo'), 'Your regex should return <code>true</code> for the string <code>"Halo"</code>.');
  - text: Your regex should return <code>false</code> for the string <code>"Hell"</code>.
    testString: assert(!myRegex.test('Hell'), 'Your regex should return <code>false</code> for the string <code>"Hell"</code>.');
  - text: You should use <code>test()</code> to test the regex.
    testString: assert(code.match(/myRegex.test\(.*\), 'You should use <code>test()</code> to test the regex.');
  - text: Your result should return <code>true</code>.
    testString: assert(result === true, 'Your result should return <code>true</code>.');

> ```

Reviewer: These tests are fine in the sense of the description, but they leave the exercise contrived — there's no need for grouping to solve this challenge with the seed string — Hello|Halo does it. If we care enough, this is solvable by asking them to match against something much longer, so the bruteforce solution would produce a very long and ugly regex.

Author: What do you suggest we ask the user to check for?

Author: This is why I was checking the regex the user had written

Author: You mean the dot with the .test()? I don't know, without the dot, the test was failing maybe something else and refer here once: https://github.com/freeCodeCamp/freeCodeCamp/blob/master/curriculum/challenges/english/02-javascript-algorithms-and-data-structures/regular-expressions/using-the-test-method.english.md that has been written with the . too. Maybe you can make a PR for that?

Reviewer: The test is failing because of something else. You can even check the results of this run to see for yourself.

Anyway, since `test` is already taught separately, why bother here? Just make this part of the seed code, so users focus on the regex.

Author: Removing the test for test then, I've made changes to the whole thing. Have a look

</section>

## Challenge Seed
<section id='challengeSeed'>

<div id='js-seed'>

```js
let myString = "Hello from the other side";
let myRegex = /False/; // Change this line

> let result = true; // Change this line

Reviewer: ```suggestion
let result = false; // Change this line
```
just so it doesn't pass its test on load.

```yml
tests:

>   - text: Your regex <code>myRegex</code> should return <code>true</code> for the string <code>"Franklin Roosevelt"</code>

Reviewer: Don't quote strings in code tags. We're programmers teaching programmers and one of the things to learn is the value of being precise, exact. With your instruction I'd try to match the quote marks as well, which is not what you're testing against.

    testString: assert(myRegex.test('Franklin Roosevelt'), 'Your regex <code>myRegex</code> should return <code>true</code> for the string <code>"Franklin Roosevelt"</code>');
  - text: Your regex <code>myRegex</code> should return <code>true</code> for the string <code>"Franklin D. Roosevelt"</code>
    testString: assert(myRegex.test('Franklin D. Roosevelt'), 'Your regex <code>myRegex</code> should return <code>true</code> for the string <code>"Franklin D. Roosevelt"</code>');
  - text: Your regex <code>myRegex</code> should return <code>true</code> for the string <code>"Eleanor Roosevelt"</code>
    testString: assert(myRegex.test('Eleanor Roosevelt'), 'Your regex <code>myRegex</code> should return <code>true</code> for the string <code>"Eleanor Roosevelt"</code>');
  - text: Your regex <code>myRegex</code> should return <code>false</code> for the string <code>"Franklin Rosevelt"</code>

>     testString: assert(!myRegex.test('Franklin Rosevelt'), 'Your regex <code>myRegex</code> should return <code>false</code> for the string <code>"Franklin Rosevelt"</code>');

Reviewer: Hmm, would it be even simpler to grasp if we just did this: "match all variants of FDR's name": with middle name, without, with just "F."? I think it would. You then use the group with ? or * to also cover the no middle name case.

Reviewer: Kinda, but to also drop Eleanor.

Author: So we have the check just for Franklin Roosevelt and other variations including his middle name?

Kind defeats the purpose, we wanted to have option grouping along with compulsory text. Like you had said: ` a(bb|cc)d` like that

Reviewer: You just add Franklin Bozo Roosevelt to the negative tests and it's covered.

Author: Doing something, check it. Also, NO, I'm not writing Franklin Bozo Roosevelt 🤣

Author: Wait, doesn't the second test already kinda do that?

It allows for any length of a middle name

Reviewer: Arbitrary ones, sure. I'll simplify it further: if you just ask to match F R and F D R, there won't be any need for an OR operator, as the regex would simply be F( D)? R. Tighter exercise.

Author: Looks fine now?

Reviewer: I guess, but I'm also tired of repeating the same comments.
